### PR TITLE
✨ Implement Data Mapper and Consent DAS Step 1 POST (#256)

### PR DESCRIPTION
**Pull request by mistryrn**
## Description of Changes
Adds router endpoints, service functions, and types necessary for Data Mapper and Consent DAS to handle the POST request for Step 1 of the Consent Wizard (Informed Consent).

Also includes a Zod package upgrade.

### Consent DAS
- Implemented POST `/participant-responses` router

### Data Mapper
- Implemented POST `/wizard/steps/informed-consent/:participantId` router

### Types
- Added `BadRequestError` to `types/httpResponses` for handling `400` bad request errors
- Added common `Date` types to `types/common` for representing and validating different date types
- Added `CreateParticipantResponse` request type to `types/consentDas/requests`
- Renamed `ParticipantResponses` request type to `ParticipantResponseArray` request in `types/consentDas/requests` to remove ambiguity with the "CreateParticipantResponse" request
- Updated `ParticipantResponse` response type to include missing data
- Added `CreateParticipantResponse` and `InformedConsent` request types to `types/dataMapper/requests`
- Updated `InformedConsent` and `ParticipantResponse` response types in `types/dataMapper/responses`

### Dependencies
- Upgraded `zod` to v3.23.3 to support new `z.string.date()` validation

### Special Instructions
🚨🚨  Note: if you haven't upgraded your `pnpm` recently, you should do that first to avoid constantly changing the lockfile version:
```bash
pnpm add -g pnpm
```
Once running the latest version of pnpm, you should install the latest dependencies and rebuild the types package:
```bash
pnpm i
pnpm build:types
```

## PR Readiness Checklist

- [x] "Expected Outcome(s)" in ticket have been met
- [x] Ticket number included in PR title
- [x] Connected ticket to PR
- [x] Labels added to PR for service name (`consent-api`, `data-mapper`, etc...), type (`chore`, `documentation`, etc...), status (`draft`, `on-hold`, etc...) **if applicable**
- [x] Manual testing completed
- [x] Builds locally without errors or warnings
- [ ] Tests are updated (if required) and passing
- [x] PR feedback has been addressed **if applicable**
- [ ] New environment variables added to `.env.schema` files, `README.md`
- [x] Added copyrights to new files

### CODE REVIEW

File: apps/consent-das/src/routers/swagger.ts [hunk 1]
```diff
@@ -84,22 +87,25 @@ const options = swaggerJsdoc({
...
-				ClinicianInviteResponse,
-				ConsentQuestionArray,
+				NanoId,
+				ParticipantResponse,
 				ParticipantResponseArray,
-				SortOrder,
 				SignatureRequest,
 				SignatureResponseArray,
-				CreateSignatureRequest,
+				SortOrder,
```
Comment (anncatton) [inline apps/consent-das/src/routers/swagger.ts:108]: thanks for ordering!

File: apps/consent-das/src/services/create.ts [hunk 2]
```diff
@@ -91,24 +92,54 @@ export const createConsentQuestion = async ({
...
+	| 'PARTICIPANT_DOES_NOT_EXIST';
+/**
+ * Creates a ParticipantResponse entry in the Consent DB
+ *
+ * @async
+ * @param {CreateParticipantResponseRequest} createParticipantResponseRequest - Participant Response data
+ * @returns {ParticipantResponse} ParticipantResponse object from Consent DB
+ */
+export const createParticipantResponse = async (
+	createParticipantResponseRequest: CreateParticipantResponseRequest,
```
Comment (anncatton) [inline apps/consent-das/src/services/create.ts:107]: it's a quibble, but i think just naming this `request` is easier to read, and because of the provided type + the function name it's clear what the request object is for

File: apps/consent-das/src/services/create.ts [hunk 3]
```diff
@@ -91,24 +92,54 @@ export const createConsentQuestion = async ({
...
+			if (error instanceof PrismaClientKnownRequestError) {
+				if (error.code === 'P2002') {
+					const errorMessage = `A participant response already exists with that '${
+						error.meta?.target ?? 'data'
+					}'`;
+					logger.error('POST /participant-responses', errorMessage, error.message);
+					return failure('PARTICIPANT_RESPONSE_EXISTS', errorMessage);
+				}
+				if (error.code === 'P2003') {
+					// Prisma error code P2003 indicates "Foreign key constraint failed"
```
Comment (anncatton) [inline apps/consent-das/src/services/create.ts:124]: Alyssa created an error handling [doc](https://github.com/OHCRN/platform/blob/develop/docs/error-handling.md#prisma-errors), there's a section for prisma errors. Let's add the `2003` type there instead of the comment here, since with the error code + the failure status there's enough info to know what's going on.

File: apps/data-mapper/src/routers/steps/informedConsent.ts [hunk 4]
```diff
@@ -109,4 +114,87 @@ router.get('/:participantId', async (req, res) => {
...
 });
 
+/**
+ * @openapi
+ * /wizard/steps/informed-consent/{participantId}:
+ *   post:
+ *     tags:
+ *       - Consent Wizard
+ *     name: Save Informed Consent
+ *     description: Save Participant's latest response for Consent Wizard - Informed Consent
```
Comment (anncatton) [inline apps/data-mapper/src/routers/steps/informedConsent.ts:124]: would reword a little `Save a new ParticipantResponse for Consent Wizard - Informed Consent`. This wording sounds a little like we're overwriting a previous response!

File: apps/data-mapper/src/routers/steps/informedConsent.ts [hunk 5]
```diff
@@ -109,4 +114,87 @@ router.get('/:participantId', async (req, res) => {
...
+ *     security:
```
Comment (anncatton) [inline apps/data-mapper/src/routers/steps/informedConsent.ts:125]: let's start removing the `bearerAuth` tags on these services

File: apps/data-mapper/src/routers/steps/informedConsent.ts [hunk 6]
```diff
@@ -109,4 +114,87 @@ router.get('/:participantId', async (req, res) => {
...
+				logger.error(
+					`GET ${ROUTER_PATH}/:participantId`,
+					'Received invalid request submitting Informed Consent',
+					requestParticipantId.error.format(),
+				);
+				return res.status(400).json(RequestValidationErrorResponse(requestParticipantId.error));
+			}
+
+			const participantResponse = await createInformedConsent({
+				informedConsentRequest: req.body,
```
Comment (anncatton) [inline apps/data-mapper/src/routers/steps/informedConsent.ts:171]: looking at this again, i like that you parsed out the args here instead of passing in the whole request, seems tidier!

File: apps/data-mapper/src/services/create.ts [hunk 7]
```diff
@@ -220,3 +229,59 @@ export const createInvite = async ({
...
+	| 'PARTICIPANT_RESPONSE_EXISTS'
+	| 'PARTICIPANT_DOES_NOT_EXIST';
+
+/**
+ * Creates an Informed Consent Participant Response entry in the Consent DAS
+ *
+ * @async
+ * @param createInformedConsentRequest - Informed Consent Participant Response request data and Participant ID
+ * @param {InformedConsentRequest} createInformedConsentRequest.informedConsentRequest - Informed Consent Participant Response request data
+ * @param {NanoId} createInformedConsentRequest.participantId - Participant ID
```
Comment (anncatton) [inline apps/data-mapper/src/services/create.ts:246]: i think a little too much documentation here 🙂 Can prob suffice with just the 2 params and their types: 
```
@param informedConsentRequest
@param participantId
@returns InformedConsentResponse
```

File: apps/data-mapper/src/services/create.ts [hunk 8]
```diff
@@ -220,3 +229,59 @@ export const createInvite = async ({
...
+ */
+export const createInformedConsent = async ({
+	informedConsentRequest,
+	participantId,
+}: {
+	informedConsentRequest: InformedConsentRequest;
+	participantId: NanoId;
+}): Promise<Result<InformedConsentResponse, CreateParticipantResponseFailureStatus>> => {
+	try {
+		const participantResponseRequest: CreateParticipantResponseRequest = {
```
Comment (anncatton) [inline apps/data-mapper/src/services/create.ts:257]: shouldn't need to assign the type to the object here. If the object is missing a necessary key/has an incorrect value type, the compiler will shout.
If you're looking to guard against unexpected keys, `satisifies` is a nice option, and you can still see more info about the obj in intellisense.

File: apps/data-mapper/src/services/create.ts [hunk 9]
```diff
@@ -220,3 +229,59 @@ export const createInvite = async ({
...
+
+		if (!parsedParticipantResponse.success) {
+			logger.error(
+				'Received invalid data in create participant response request response',
+				parsedParticipantResponse.error.issues,
+			);
+			return failure('SYSTEM_ERROR', parsedParticipantResponse.error.message);
+		}
+
+		const informedConsent: InformedConsentResponse = {
```
Comment (anncatton) [inline apps/data-mapper/src/services/create.ts:278]: same comment as above, i don't think the extra typing for this is necessary, esp since at this part of the flow we've run `safeParse` on the response.

File: apps/data-mapper/src/services/das/consent.ts [hunk 10]
```diff
@@ -244,6 +246,50 @@ export const createParticipantConsentData = async (
...
 
+/**
+ * Makes request to Consent DAS to record a new Participant Response
+ *
+ * @async
+ * @param {CreateParticipantResponseRequest} createParticipantResponseRequest - Participant Response data
+ * @returns {ParticipantResponse} ParticipantResponse object stored in Consent DB
+ */
+export const createParticipantResponse = async (
+	createParticipantResponseRequest: CreateParticipantResponseRequest,
```
Comment (anncatton) [inline apps/data-mapper/src/services/das/consent.ts:257]: similar comment to above, i feel like the naming is more detailed than necessary, could just be `request: CreateParticipantResponseRequest`. The typing is very clear about what the request is.
I know we discussed this a little earlier, but looking at it again it just feels slightly cluttered to my eyes!

File: packages/types/src/common/Date.ts [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+// For YYYY-MM-DD Date strings (i.e. '2021-01-01')
+export const DateString = z.string().date();
+export type DateString = z.infer<typeof DateString>;
+
+// For ISO8601 Datetime strings (i.e. '2021-01-01T00:00:00.000Z')
+export const DateTime = z.string().datetime();
+export type DateTime = z.infer<typeof DateTime>;
+
+// For Datetime strings with timezone offset (i.e. '2021-01-01T00:00:00.000+00:00')
```
Comment (anncatton) [inline packages/types/src/common/Date.ts:34]: Thanks for the comments on these! Maybe we can create a followup ticket just to look over our other uses of dates and any types we might need to update now that this is available in Zod? (Appreciate that you did not try to include that in this PR! 😛 )

File: apps/consent-das/src/routers/participantResponses.ts [hunk 12]
```diff
@@ -136,39 +142,52 @@ router.get('/:participantId/:consentQuestionId', async (req, res) => {
  *     security:
  *       - bearerAuth: []
```
Comment (anncatton) [inline apps/consent-das/src/routers/participantResponses.ts:143]: can remove `bearerAuth`

**Review by anncatton — COMMENTED** [review]
Looks awesome! You're gonna love that most of my comments are..."there is too much typing here" 😂

File: apps/data-mapper/src/routers/steps/informedConsent.ts (hunk 5 above)
Comment (mistryrn) [reply apps/data-mapper/src/routers/steps/informedConsent.ts:125]: Done ✅

File: apps/data-mapper/src/routers/steps/informedConsent.ts (hunk 4 above)
Comment (mistryrn) [reply apps/data-mapper/src/routers/steps/informedConsent.ts:124]: Good call, that sounds much better. Updated 🙂

File: apps/data-mapper/src/services/create.ts (hunk 7 above)
Comment (mistryrn) [reply apps/data-mapper/src/services/create.ts:246]: We can do this, but it will create a warning "no param with that name" on `participantId` if we treat it as separate params instead of a single object param, because this function is taking a single object with two named keys:
Screenshot 2024-04-25 at 9 24 07 AM

According to the docs (https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#param-and-returns), this is the recommended way to document params for keys of an object:
Screenshot 2024-04-25 at 9 22 40 AM

I can change the function to accept two separate params instead of a single object if we'd prefer to simplify the JSDocs here. Let me know which direction you'd like to go 🙂

File: apps/data-mapper/src/services/create.ts (hunk 8 above)
Comment (mistryrn) [reply apps/data-mapper/src/services/create.ts:257]: Updated

File: apps/data-mapper/src/services/create.ts (hunk 9 above)
Comment (mistryrn) [reply apps/data-mapper/src/services/create.ts:278]: Updated

File: apps/data-mapper/src/services/das/consent.ts (hunk 10 above)
Comment (mistryrn) [reply apps/data-mapper/src/services/das/consent.ts:257]: Updated!

File: packages/types/src/common/Date.ts (hunk 11 above)
Comment (mistryrn) [reply packages/types/src/common/Date.ts:34]: Thanks for reminding me, ticket created here: https://github.com/OHCRN/platform/issues/573

**Review by mistryrn — COMMENTED** [review]
Thanks for the feedback @anncatton ! PR updated to address everything except for one note about the JSDoc params – left a question in there for you before I make any changes there 🙂

File: apps/data-mapper/src/services/create.ts (hunk 7 above)
Comment (anncatton) [reply apps/data-mapper/src/services/create.ts:246]: since we just chatted IRL but just adding here for posterity...we can do the non-object params instead and rely on the types to yell if an unexpected parameter is passed. The recommended way just looks so crowded!

File: docs/error-handling.md [hunk 13]
```diff
@@ -48,21 +48,27 @@ We have an enum in the [`ErrorResponse.ts`](../packages/types/src/httpResponses/
 
 ```ts
 export enum ErrorName {
-	SERVER_ERROR = 'ServerError',
+	BAD_REQUEST_ERROR = 'BadRequestError',
```
Comment (anncatton) [inline docs/error-handling.md:51]: thanks for updating this too!

File: docs/error-handling.md [hunk 14]
```diff
@@ -148,4 +154,5 @@ When implementing services in the DAS that query the Prisma client, it's helpful
 | Prisma Client Code                                                                | Message                                                                                                   |
 | --------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------- |
 | [P2002](https://www.prisma.io/docs/reference/api-reference/error-reference#p2002) | "Unique constraint failed on the {constraint}”                                                            |
+| [P2003](https://www.prisma.io/docs/reference/api-reference/error-reference#p2003) | "Foreign key constraint failed”                                                                           |
```
Comment (anncatton) [inline docs/error-handling.md:157]: 🙏

**Review by anncatton — APPROVED** [review]
Lovely! 🎨
